System.Containers/ContainerHelpers.cs

```diff
@@ -0,0 +1,92 @@
+namespace System.Containers;
+
+using System.Diagnostics.CodeAnalysis;
+using System.Text.RegularExpressions;
+
+public static class ContainerHelpers
+{
+    private static string[] knownImageSeparators = { "-", ".", "_" };
+
+    private static Regex imageTagRegex = new Regex("[a-zA-Z0-9_][a-zA-Z0-9._-]{0,127}");
+
+    private static Regex imageNameRegex = new Regex("[a-z0-9]+([._-][a-z0-9]+)*(/[a-z0-9]+([._-][a-z0-9]+)*)*");
+
+    /// <summary>
+    /// Given some "fully qualified" image name (e.g. mcr.microsoft.com/dotnet/runtime), return
+    /// a valid UriBuilder. This means appending 'https' if the URI is not absolute, otherwise UriBuilder will throw.
+    /// </summary>
+    /// <param name="containerBase"></param>
+    /// <returns>A UriBuilder with the given containerBase, or, if containerBase is relative, https:// + containerBase</returns>
+    private static Uri? ContainerImageToUri(string containerBase)
+    {
+        Uri uri = new Uri(containerBase, UriKind.RelativeOrAbsolute);
+
+        try
+        {
+            return uri.IsAbsoluteUri ? uri : new Uri(containerBase.Contains("localhost") ? "http://" : "https://" + uri);
+        }
+        catch (Exception e)
+        {
+            Console.WriteLine("Failed parsing the container image into a UriBuilder: {0}", e);
+            return null;
+        }
+    }
+
+    /// <summary>
+    /// Ensures the given image name is valid.
+    /// Spec: https://github.com/opencontainers/distribution-spec/blob/4ab4752c3b86a926d7e5da84de64cbbdcc18d313/spec.md#pulling-manifests
+    /// </summary>
+    /// <param name="imageName"></param>
+    /// <returns></returns>
+    public static bool IsValidImageName(string imageName)
+    {
+        return imageNameRegex.IsMatch(imageName);
+    }
+
+    /// <summary>
+    /// Ensures the given tag is valid.
+    /// Spec: https://github.com/opencontainers/distribution-spec/blob/4ab4752c3b86a926d7e5da84de64cbbdcc18d313/spec.md#pulling-manifests
+    /// </summary>
+    /// <param name="imageTag"></param>
+    /// <returns></returns>
+    public static bool IsValidImageTag(string imageTag)
+    {
+        return imageTag.Length <= 128 && imageTagRegex.IsMatch(imageTag);
```

Comment:
The length check is in the regex, but `IsMatch` means "does the regex match any substring?" so it's not applying. Can the regexes be `^...$`?